librad/src/git/types/reference.rs

```diff
@@ -0,0 +1,280 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    fmt::{self, Display},
+    marker::PhantomData,
+};
+
+use crate::{git::ext, hash::Hash, uri::RadUrn};
+
+use super::{existential::SomeNamespace, Force, Refspec, SymbolicRef};
+
+/// Type witness for a [`Reference`] that should point to a single reference.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Single;
+
+/// Type witness for a [`Reference`] that should point to multiple references.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Multiple;
+
+pub type Namespace = Hash;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RefsCategory {
+    Heads,
+    Rad,
+}
+
+impl Display for RefsCategory {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::Heads => f.write_str("heads"),
+            Self::Rad => f.write_str("rad"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Reference<Namespaced, Remote, Cardinality> {
+    /// The remote portion of this reference.
+    pub remote: Option<Remote>,
+    /// Where this reference falls under, i.e. `rad` or `heads`.
+    pub category: RefsCategory,
+    /// The path of the reference, e.g. `feature/123`, `dev`.
+    pub name: String,
+
+    pub(super) _namespace: Namespaced,
+    pub(super) _cardinality: PhantomData<Cardinality>,
+}
+
+// Polymorphic definitions
+impl<Namespaced: Clone, R: Clone, N: Clone> Reference<Namespaced, R, N> {
+    /// Set the remote portion of thise reference.
+    ///
+    /// Note: This is consuming.
```

Comment:
Uh-oh, idiotmatic polizei instructed me: `set_x` is `&self -> ()`, `with_x` is `self -> Self`, and builder-style `x` is `&mut self -> &mut Self` :man_shrugging:

Reply:
Well, they should have apprehended you because I didn't change the naming just the location :stuck_out_tongue_closed_eyes: 
I'll change it while I'm shaving yaks :water_buffalo:

Reply:
Ouch :handcuffs: